conformance/tests/httproute-request-percentage-mirror.go

```diff
@@ -0,0 +1,234 @@
+					},
+				},
+				Backend: "infra-backend-v1",
+				MirroredTo: []http.MirroredBackend{
+					{
+						BackendRef: http.BackendRef{
+							Name:      "infra-backend-v2",
+							Namespace: ns,
+						},
+						Percent: ptr.To(int32(20)),
+					},
+				},
+			}, {
+				Request:   http.Request{Path: "/percent-mirror-fraction"},
+				Namespace: ns,
+				ExpectedRequest: &http.ExpectedRequest{
+					Request: http.Request{
+						Path: "/percent-mirror-fraction",
+					},
+				},
+				Backend: "infra-backend-v1",
+				MirroredTo: []http.MirroredBackend{
+					{
+						BackendRef: http.BackendRef{
+							Name:      "infra-backend-v2",
+							Namespace: ns,
+						},
+						Percent: ptr.To(int32(50)),
+					},
+				},
+			}, {
+				Request: http.Request{
+					Path: "/percent-mirror-and-modify-headers",
+					Headers: map[string]string{
+						"X-Header-Remove":     "remove-val",
+						"X-Header-Add-Append": "append-val-1",
+					},
+				},
+				ExpectedRequest: &http.ExpectedRequest{
+					Request: http.Request{
+						Path: "/percent-mirror-and-modify-headers",
+						Headers: map[string]string{
+							"X-Header-Add":        "header-val-1",
+							"X-Header-Add-Append": "append-val-1,header-val-2",
+							"X-Header-Set":        "set-overwrites-values",
+						},
+					},
+					AbsentHeaders: []string{"X-Header-Remove"},
+				},
+				Namespace: ns,
+				Backend:   "infra-backend-v1",
+				MirroredTo: []http.MirroredBackend{
+					{
+						BackendRef: http.BackendRef{
+							Name:      "infra-backend-v2",
+							Namespace: ns,
+						},
+						Percent: ptr.To(int32(35)),
+					},
+				},
+			},
+		}
+
+		for i := range testCases {
+			expected := testCases[i]
+			t.Run(expected.GetTestCaseName(i), func(t *testing.T) {
+				// Assert request succeeds before doing our distribution check
+				http.MakeRequestAndExpectEventuallyConsistentResponse(t, suite.RoundTripper, suite.TimeoutConfig, gwAddr, expected)
+
+				// Override to not have more requests than expected
+				suite.TimeoutConfig.RequiredConsecutiveSuccesses = 1
+				// used to limit number of parallel go routines
+				semaphore := make(chan struct{}, concurrentRequests)
+				var wg sync.WaitGroup
+
+				for k := 0; k < numDistributionChecks; k++ {
+					timeNow := time.Now()
+					for j := 0; j < totalRequests; j++ {
+						wg.Add(1)
+						semaphore <- struct{}{}
+						go func(t *testing.T, r roundtripper.RoundTripper, timeoutConfig config.TimeoutConfig, gwAddr string, expected http.ExpectedResponse) {
+							defer wg.Done()
+							defer func() { <-semaphore }()
+							http.MakeRequestAndExpectEventuallyConsistentResponse(t, r, timeoutConfig, gwAddr, expected)
+						}(t, suite.RoundTripper, suite.TimeoutConfig, gwAddr, expected)
+					}
+					wg.Wait()
+					if err := testMirroredRequestsDistribution(t, suite, expected, timeNow); err != nil {
+						t.Logf("Traffic distribution test failed (%d/%d): %s", k+1, numDistributionChecks, err)
+						time.Sleep(2 * time.Second)
```

Comment:
Why do we need such a sleep here?

Reply:
The reason we sleep is that we pass timeNow to the function that reads the logs and counts how many request were mirrored - and if I dont sleep the time overlaps and we get more logs read than we expected.